src/comlink.ts

```diff
@@ -371,6 +375,39 @@ function throwIfProxyReleased(isReleased: boolean) {
   }
 }
 
+function releaseEndpoint(ep: Endpoint) {
+  return requestResponseMessage(ep, {
+    type: MessageType.RELEASE,
+  }).then(() => {
+    closeEndPoint(ep);
+  });
+}
+
+interface FinalizationRegistry<T> {
+  new (cb: (heldValue: T) => void): FinalizationRegistry<T>;
+  register(weakItem: any, heldValue: T, unregisterToken: any): void;
+}
+declare var FinalizationRegistry: FinalizationRegistry<Endpoint>;
+
+const proxyCounter = new WeakMap<Endpoint, number>();
+const proxyFinalizers =
+  "FinalizationRegistry" in self &&
+  new FinalizationRegistry((ep: Endpoint) => {
+    const newCount = (proxyCounter.get(ep) || 0) - 1;
+    proxyCounter.set(ep, newCount);
+    if (newCount === 0) {
+      releaseEndpoint(ep);
+    }
+  });
+
+function registerProxy(proxy: any, ep: Endpoint) {
+  const newCount = (proxyCounter.get(ep) || 0) + 1;
+  proxyCounter.set(ep, newCount);
+  if (proxyFinalizers) {
+    proxyFinalizers.register(proxy, ep, ep);
```

Comment:
Shouldn't `proxy` be the `unregisterToken` so that a manually invocation of `releaseProxy` could call `proxyFinalizers.unregister(proxy)` (and decrease the endpoint count)?

I made sure this was a supported use case hence the note:
> For example, registering an object with itself as its unregister token would not keep the object alive forever.

Reply:
I just put _something_ for the `unregisterToken`, because I am not using `unregister()` and I wasn’t sure if `undefined` would cause problems. But you are totally right that I should be calling `unregister()` when `releaseProxy()` is called manually! Thanks for taking the time to look at this.